### PR TITLE
Add trigger functionality

### PR DESCRIPTION
**Pull request by shadowwa**
This commit add:
 - adaptative trigger configuration, you can set preset mode (feedback, weapon, bow…) with parameter
 - firmware inf
 - speaker configuration (switch speaker/header, change volume)
 - rumble motor attenuation

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -42,3 +56,11 @@ Also installed by Steam, so you may already have it configured. If not, create `
 
     # PS5 DualSense controller over bluetooth hidraw
     KERNEL=="hidraw*", KERNELS=="*054C:0CE6*", MODE="0660", TAG+="uaccess"
+
+### Credit
+
+nowrep for the original dualsensectl code
```
Comment (nowrep) [inline README.md:62]: Remove this please

File: main.c [hunk 2]
```diff
@@ -51,21 +51,65 @@
...
 
+#define DS_TRIGGER_EFFECT_OFF 0x05
+#define DS_TRIGGER_EFFECT_FEEDBACK 0x21
+#define DS_TRIGGER_EFFECT_BOW 0x22
+#define DS_TRIGGER_EFFECT_GALLOPING 0x23
+#define DS_TRIGGER_EFFECT_WEAPON 0x25
+#define DS_TRIGGER_EFFECT_VIBRATION 0x26
+#define DS_TRIGGER_EFFECT_MACHINE 0x27
+
+// debug
```
Comment (nowrep) [inline main.c:100]: Remove this

File: main.c [hunk 3]
```diff
@@ -567,6 +682,341 @@ static int command_microphone_led(struct dualsense *ds, char *state)
...
+
+    if (!strcmp(trigger, "right") || !strcmp(trigger, "both")) {
+        rp.common->valid_flag0 = DS_OUTPUT_VALID_FLAG0_RIGHT_TRIGGER_MOTOR_ENABLE;
+    }
+    if (!strcmp(trigger, "left") || !strcmp(trigger, "both")) {
+        rp.common->valid_flag0 |= DS_OUTPUT_VALID_FLAG0_LEFT_TRIGGER_MOTOR_ENABLE;
+    }
+
+     /*
+    printf("validflag0 value: %03i %#04x : "BYTE_TO_BINARY_PATTERN"\n", rp.common->valid_flag0,  rp.common->valid_flag0,BYTE_TO_BINARY(rp.common->valid_flag0));
```
Comment (nowrep) [inline main.c:773]: Remove

File: main.c [hunk 4]
```diff
@@ -567,6 +682,341 @@ static int command_microphone_led(struct dualsense *ds, char *state)
...
+        fprintf(stderr, "amplitude must be between 1 and 8\n");
+        return 1;
+    }
+     if (!(frequency > 0)) {
+        fprintf(stderr, "frequency must be greater than 0\n");
+        return 1;
+    }
+
+    uint8_t strength_array[10] = {0};
+    for (int i = position; i < 10; i++)
```
Comment (nowrep) [inline main.c:1001]:
```c
for (int i = position; i < 10; i++) {
    strength_array[i] = amplitude;
}
```

**Review by nowrep — CHANGES_REQUESTED** [review]
Thanks for the PR.

It's mostly nitpicks and it should be good to go.
Also please cleanup the commits (there should be no merge commits etc), otherwise I'll squash it as single commit.

File: README.md (hunk 1 above)
Comment (shadowwa) [reply README.md:62]: do you want me to remove the whole credit paragraph or only the line 62 ?

Comment (nowrep) [reply README.md:62]: Remove the line 62 (crediting me), you can keep the rest if you want.

**Comment by shadowwa** [comment]
I've applied the requested changes.
for the commits cleanup, I'll try to play with rebase later, bu I don't mind if you squash the whole thing

**Comment by shadowwa** [comment]
I think it's all good now. The commits are cleaned up and I fixed some other coding style I've found in the process.

**Review by nowrep — APPROVED** [review]

**Comment by nowrep** [comment]
Thanks!

**Pull request merged**